src/D2L.Bmx/Aws/AwsCredsCache.cs

```diff
@@ -0,0 +1,103 @@
+}
+
+internal class AwsCredsCache : IAwsCredentialCache {
+	private static void WriteTextToFile( string path, string content ) {
+		var op = new FileStreamOptions {
+			Mode = FileMode.Create,
+			Access = FileAccess.Write,
+		};
+		if( !RuntimeInformation.IsOSPlatform( OSPlatform.Windows ) ) {
+			op.UnixCreateMode = UnixFileMode.UserRead | UnixFileMode.UserWrite;
+		}
+		using var writer = new StreamWriter( path, op );
+		writer.Write( content );
+	}
+
+	private List<AwsCacheModel> GetAllCache() {
+		if( !File.Exists( BmxPaths.AWS_CREDS_CACHE_FILE_NAME ) ) {
+			return new();
+		}
+		try {
+			string sessionsJson = File.ReadAllText( BmxPaths.AWS_CREDS_CACHE_FILE_NAME );
+			return JsonSerializer.Deserialize( sessionsJson, SourceGenerationContext.Default.ListAwsCacheModel )
+				?? new();
+		} catch( JsonException ) {
+			return new();
+		}
+	}
+
+	void IAwsCredentialCache.SetCredentials( string org, string user, AwsRole role, AwsCredentials credentials ) {
+		var ttlLongerThan = DateTime.UtcNow.AddMinutes( 10 );
+
+		var allEntries = GetAllCache()
+		// Remove expired/expiring & entries not related to current user
+		.Where( o =>
+			o.Credentials.Expiration >= ttlLongerThan
+			&& o.User == user
+			&& o.Org == org
+		)
+		.ToList();
+
+		allEntries.Add( new AwsCacheModel(
+			Org: org,
+			User: user,
+			RoleArn: Arn.Parse( role.RoleArn ),
+			Credentials: credentials
+		) );
+
+		EnsureUniqueRoleNames( allEntries );
+
+		var prunedEntries = allEntries
+			// Prune older (closer to expiry) credentials for the same role
+			.GroupBy( o => o.RoleArn, ( _, entries ) =>
+				entries.OrderBy( o => o.Credentials.Expiration )
+			)
+			.Select( o => o.Last() );
+
+		string jsonString = JsonSerializer.Serialize( prunedEntries.ToList(),
+			SourceGenerationContext.Default.ListAwsCacheModel );
+
+		WriteTextToFile( BmxPaths.AWS_CREDS_CACHE_FILE_NAME, jsonString );
+	}
+
+	AwsCredentials? IAwsCredentialCache.GetCredentials(
+		string org,
+		string user,
+		string roleName,
+		int duration
+	) {
+		List<AwsCacheModel> allEntries = GetAllCache();
+
+		// Too stale, within a window (of 10 mins) if requested duration was <= 20 mins, 15 mins otherwise
+		// Avoids over-eagerly invalidating cache when requesting AWS credentials for short durations
+		var ttlLongerThan = DateTime.UtcNow.AddMinutes( duration > 20 ? 15 : duration - 5 );
+		var matchedEntries = allEntries.FindAll(
+			o => o.Org == org
+			&& o.User == user
+			&& o.RoleArn.Resource.Split( "/" ).Last().Equals( roleName, StringComparison.OrdinalIgnoreCase )
+			&& o.Credentials.Expiration >= ttlLongerThan
+		);
+
+		EnsureUniqueRoleNames( matchedEntries );
+
+		return matchedEntries.FirstOrDefault()?.Credentials;
+	}
+
+	// Since we retrieve credentials without talking to APIs we are using RoleName as a cache key,
+	// Since we do not have account ID we are at risk of potentially retrieving credentials for the
+	// the wrong role (since role names aren't globally unique), guard against this
```

Comment:
should/can we cache by (normalized) account name + role name, and relax this restriction?

("normalized" because we don't require case sensitivity on account name)

Reply:
I was writing an explanation for this and realised, if we put accountName in cache key, then we can relax this requirement again. Since at least in theory the alias has to be unique?

Otherwise selecting account via BMX would not work correctly (since we match account alias to Okta App label).

Reply:
it's actually the "app name" in Okta and has nothing to do with account alias. but anyway, yes it should be a unique combination within an org if bmx works at all.

Reply:
I opted to call the additional cache key "accountName" at least throughout the rest of BMX we call it by account name when in practice it relates to name in Okta which in theory is unrelated to AWS Account alias (would make for a very confusing Okta App chiclet at least as far as D2Ler's are used to).